modules/gdscript/editor/gdscript_docgen.h

```diff
@@ -0,0 +1,42 @@
+/**************************************************************************/
+/*  gdscript_docgen.h                                                     */
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef GDSCRIPT_DOCGEN_H
+#define GDSCRIPT_DOCGEN_H
+
+#include "../gdscript_parser.h"
+#include "core/doc_data.h"
+
+class GDScriptDocGen {
+public:
+	static void generate_docs(GDScript *p_script, const GDScriptParser::ClassNode *p_class, bool p_inner = false);
+};
```

Comment:
Does it need to be a class? Can be just a standalone method, we do this sometimes.

Edit: Though I can see that you made it a friend of other GDScript classes, so perhaps it needs scope access? Then I guess it's okay as a class.

Reply:
I was trying to remember why I did this and I think I remember :) I'm sure it's a common C++ problem, but this was the only solution I could find but maybe there are better ones!

Here's a rundown of why I think this is necessary :)

1. This method needs access to the `ClassNode` in order to use the analyzer's datatypes. This means that wherever this method is declared needs to include `gdscript_parser.h`. Same problem if we store the `ClassNode` as a member of `GDScript` to avoid having it in the method signature.
2. If we make this a `GDScript` class method, then in order for the method to have this signature, the `gdscript.h` file needs to include `gdscript_parser.h`, which includes `gdscript_cache.h`, which includes `gdscript.h`, which is a circular include loop.
3. If we make this into a static method inside the `gdscript.cpp` file, we no longer have an include loop, but now we also don't have access to the `GDScript` class' documentation-related internals, so most of the `p_script->...` accesses would no longer work.
4. We inline the entire docgen code in `GDScript::reload()` 🤐 

Is there another way?

Reply:
I'm not sure if there is another, more preferred way, but I like your justification :)